File: 1305.两颗二叉搜索树中的所有元素/getAllElements.go
package problem1305

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// 这道题 = 中序遍历 + 归并排序
func getAllElements(root1 *TreeNode, root2 *TreeNode) []int {
	var (
		result     []int
		list1      []int
		list2      []int
		inorderDFS func(node *TreeNode, container *[]int)
	)
	inorderDFS = func(node *TreeNode, container *[]int) {
		if node == nil {
			return
		}
		inorderDFS(node.Left, container)
		*container = append(*container, node.Val)
		inorderDFS(node.Right, container)
	}
	inorderDFS(root1, &list1)
	inorderDFS(root2, &list2)
	// 归并
	i, j, len1, len2 := 0, 0, len(list1), len(list2)
	for i < len1 && j < len2 {
		if list1[i] <= list2[j] {
			result = append(result, list1[i])
			i++
		} else {
			result = append(result, list2[j])
			j++
		}
	}
	for i < len1 {
		result = append(result, list1[i])
		i++
	}
	for j < len2 {
		result = append(result, list2[j])
		j++
	}
	return result
}

File: 1305.两颗二叉搜索树中的所有元素/getAllElements.js
// 显然是中序遍历 + 归并排序
function getAllElements(root1, root2) {
	const result = []
	const list1 = []
	const list2 = []
	helper(root1, list1)
	helper(root2, list2)

	let i = 0, j = 0
	while (i < list1.length && j < list2.length) {
		if (list1[i] < list2[j]) {
			result.push(list1[i++])
		} else {
			result.push(list2[j++])
		}
	}
	while (i < list1.length) {
		result.push(list1[i++])
	}
	while (j < list2.length) {
		result.push(list2[j++])
	}
	return result
}

function helper(root, list) {
	if (!root) return
	helper(root.left, list)
	list.push(root.val)
	helper(root.right, list)
}

// 迭代器版本，leetcode173
// 参考 https://leetcode-cn.com/problems/all-elements-in-two-binary-search-trees/solution/gou-jian-bstdie-dai-qi-lai-huo-de-geng-h-ib0l/
class BSTIterator {
	constructor(root) {
		this.stack = []
		this.pushToStack(root)
	}
	next() {
		const cur = this.stack.pop()
		this.pushToStack(cur.right)
		return cur.val
	}
	peek() {
		return this.stack[this.stack.length - 1].val
	}
	hasNext() {
		return this.stack.length
	}
	pushToStack(root) {
		while (root) {
			this.stack.push(root)
			root = root.left
		}
	}
}

function getAllElements1(root1, root2) {
	const res = []
	const r1 = new BSTIterator(root1)
	const r2 = new BSTIterator(root2)

	while (r1.hasNext() && r2.hasNext()) {
		if (r1.peek() > r2.peek()) {
			res.push(r2.next())
		} else {
			res.push(r1.next())
		}
	}
	while (r1.hasNext()) res.push(r1.next())
	while (r2.hasNext()) res.push(r2.next())

	return res
}
